gRPC reflection security (#33092)
dotnet/AspNetCore.Docs: aspnetcore/grpc/test-tools.md

* Client apps that support gRPC reflection can call the reflection service to discover services hosted by the server.
* gRPC services are still called from the client. Reflection only enables service discovery and doesn't bypass server-side security. Endpoints protected by [authentication and authorization](xref:grpc/authn-and-authz) require the caller to pass credentials for the endpoint to be called successfully.

### Reflection service security

> gRPC reflection returns a list of available APIs, which could be sensitive information. Care should be taken to limit access to the gRPC reflection service.

Reviewer: ```suggestion
gRPC reflection returns a list of available APIs, which could contain sensitive information. Care should be taken to limit access to the gRPC reflection service.
```

> gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.Aspnetcore.Hosting.HostingEnvironmentExtensions.Isdevelopment> returns true:

Author: ```suggestion
gRPC reflection is usually only required in a local development environment. For local development, the reflection service should only be mapped when [IsDevelopment](/en-us/dotnet/api/microsoft.aspnetcore.hosting.hostingenvironmentextensions.isdevelopment) returns true:
```

> gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:

Reviewer: ```suggestion
gRPC reflection is usually only required in a local development environment. For local development, the reflection service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:
```

> gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:

Reviewer 2: ```suggestion
gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.Aspnetcore.Hosting.HostingEnvironmentExtensions.Isdevelopment> returns true:
```

> gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:

Reviewer: ```suggestion
gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when [IsDevelopment](/en-us/dotnet/api/microsoft.aspnetcore.hosting.hostingenvironmentextensions.isdevelopment) returns true:
```

Reviewer 2: Looks like Rick and I are in at same time, I will get out.

```csharp
if (env.IsDevelopment())
{
    app.MapGrpcReflectionService();
}
```

> Access to the service can be controlled through standard ASP.NET Core authorization extension methods, such as <xref:Microsoft.AspNetCore.Builder.AuthorizationEndpointConventionBuilderExtensions.AllowAnonymous<TBuilder>(TBuilder)> and <xref:Microsoft.AspNetCore.Builder.AuthorizationEndpointConventionBuilderExtensions.RequireAuthorization<TBuilder>(TBuilder)>.

Reviewer: ```suggestion
Access to the service can be controlled through standard ASP.NET Core authorization extension methods, such as [`AllowAnonymous<TBuilder>`](/dotnet/api/microsoft.aspnetcore.builder.authorizationendpointconventionbuilderextensions.allowanonymous) and [`RequireAuthorization<TBuilder>`](/dotnet/api/microsoft.aspnetcore.builder.authorizationendpointconventionbuilderextensions.requireauthorization).
```

> For example, if an app has been configured to require authorizaiton by default, configuration the gRPC reflection endpoint with `AllowAnonymous()` to skip authentication and authorization.

Reviewer: ```suggestion
For example, if an app has been configured to require authorization by default, configuration the gRPC reflection endpoint with `AllowAnonymous` to skip authentication and authorization.
```
